### ISSUE TITLE
iquest fails after 2 hours

### ISSUE THREAD

**Issue by mcast** [irods/irods#5852]
### iRODS Version, OS and Version

iRODS 4.2.7 on Ubuntu Bionic 18.04.4, running a copied-isolated-migrated iCAT with no real iRES storage.

iCAT on Postgres, more info available.

### What did you try to do?

I was exercising the copy-iCAT on a Postgres backend, prior to migration of our live iCAT. This performance test is about 45 minutes, running `iquest`s in two forms,

* `select min(DATA_ID), max(DATA_ID), count(DATA_ID) where META_DATA_ATTR_NAME = 'id_ifile'` / `and DATA_ID between '22389155' '37041407'`; background info in #4855
* `select ORDER(DATA_ID), ORDER(DATA_REPL_NUM), DATA_REPL_STATUS, COLL_NAME, DATA_NAME, DATA_CHECKSUM, DATA_SIZE, DATA_RESC_HIER, DATA_PATH, META
_DATA_ATTR_VALUE where META_DATA_ATTR_NAME = 'id_ifile' and DATA_ID between '51235985' '51693660'` repeatedly on windows of ~100k files

with some quicker resource metadata lookups

* `select RESC_ID, RESC_NAME, RESC_TYPE_NAME, RESC_PARENT where RESC_ZONE_NAME = 'cgp'`
* `select RESC_NAME, RESC_VAULT_PATH, RESC
_CLASS_NAME, RESC_LOC, RESC_STATUS where RESC_TYPE_NAME like 'unix%file%system' and RESC_ZONE_NAME = 'cgp'`

I was looking for performance problems (didn't find any), but early on there was one slow failure. So I ran the performance test repeatedly and got another slow failure.

### Expected behavior

Except for two failures, the 5300+ other `iquest`s among these tests have all completed in under 60 seconds, 99.8% under 17 seconds.

These are big-ish queries because I'm traversing the entire listing of 17.3E6 file replicas, but they are designed not to be monsters which knock things over.

### Observed behavior (including steps to reproduce, if applicable)

So far the only way to reproduce this is to run the thing repeatedly and occasionally it breaks.  Our DBA reports that similar failures have been seen when the (non-iRODS?) client is in our Openstack cluster - and it has always been the client.

The second iquest failure:
```
:run(2021-09-15 05:56:40.833744) iquest  --no-page  -z  /cgp  '%s %s %s %s %s %s %s %s %s %s'  "select ORDER(DATA_ID), ORDER(DATA_REPL_NUM), DATA_REPL_STATUS, COLL_NAME, DATA_NAME, DATA_CHECKSUM, DATA_SIZE, DATA_RESC_HIER, DATA_PATH, META_DATA_ATTR_VALUE where META_DATA_ATTR_NAME = 'id_ifile' and DATA_ID between '23876440' '23990845'"; [7238.631s]
 ERROR: iquest Error: queryAndShowStrCond failed status = -806000 CAT_SQL_ERR
```
That's a start time in UTC (localtime is BST, we forgot to tell the server); plus ~two hours running, making the query 06:56:40 to 08:57:18 BST. `iquest` format params massaged for clarity.

Our DBA reports (whitespace is probably mangled, username translated for context):

> The database cluster has been running normally the entire time.  The cluster log file does contain these entries:
> 
> ```
> 2021-09-15 07:12:26.093 BST [2007777] irods@irodscgp LOG:  could not send data to client: Connection timed out
> 2021-09-15 07:12:26.093 BST [2007777] irods@irodscgp STATEMENT:  select distinct R_DATA_MAIN.data_id ,R_DATA_MAIN.data_repl_num ,R_DATA_MAIN.data_is_dirty ,R_COLL_MAIN.coll_name ,R_DATA_MAIN.data_name ,R_DATA_MAIN.data_checksum ,R_DATA_MAIN.data_size ,R_DATA_MAIN.resc_id ,R_DATA_MAIN.data_path ,r_data_meta_main.meta_attr_value  from  R_DATA_MAIN , R_COLL_MAIN , R_META_MAIN r_data_meta_main , R_OBJT_METAMAP r_data_metamap  where r_data_meta_main.meta_attr_name = $1  AND R_DATA_MAIN.data_id BETWEEN $2 AND $3  AND r_data_metamap.meta_id = r_data_meta_main.meta_id  AND R_DATA_MAIN.data_id = r_data_metamap.object_id  AND R_COLL_MAIN.coll_id = R_DATA_MAIN.coll_id  order by R_DATA_MAIN.data_id, R_DATA_MAIN.data_repl_num, R_COLL_MAIN.coll_name, R_DATA_MAIN.data_name, R_DATA_MAIN.data_repl_num
> 2021-09-15 07:12:26.093 BST [2007777] irods@irodscgp FATAL:  connection to client lost
> ```
> 
> We don’t have the cluster configured to record when connections are started, so I can't tell you where that connection came from, but I’m guessing that @kript can trawl the iRODS server logs to match the time and figure out which client host was involved.
> 
> However, the connection was dropped at the client end, not the database!

so it looks like the client fell over after 15m 46s waiting for an answer that otherwise reliably returns inside a minute.

There have been ongoing upgrade works on the Openstack cluster hosting the copy-iCAT, but we believe there would have only been one interruption to networking on the copy-iCAT host and that was an _hour (hmm, timezone problem?) before_ the first failure.

**Comment by mcast** [comment]
More about the first failure,

```
:run(2021-09-13 09:35:14.292242) iquest  --no-page  -z  /cgp  %s %s %s  select min(DATA_ID), max(DATA_ID), count(DATA_ID) where META_DATA_ATTR_NA
ME = 'id_ifile' and DATA_ID between '37041408' '51693660'; [7279.382s]                                                                                         
ERROR: iquest Error: queryAndShowStrCond failed status = -806000 CAT_SQL_ERR                                                                                
```
Again UTC, so it ran 10:35:14 to 12:36:33 BST.

@kript reported

> On the server end I saw;
>
> ```
> Sep 13 11:36:33 pid:6388 NOTICE: bindVar[1]=id_ifile
> Sep 13 11:36:33 pid:6388 NOTICE: bindVar[2]=37041408
> Sep 13 11:36:33 pid:6388 NOTICE: bindVar[3]=51693660
> Sep 13 11:36:33 pid:6388 NOTICE: cllExecSqlWithResult: SQLExecDirect error: -1, sql:select distinct min(R_DATA_MAIN.data_id ) ,max(R_DATA_MAIN.data_id ) ,count(R_DATA_MAIN.data_id )  from  R_DATA_MAIN , R_META_MAIN r_data_meta_main , R_OBJT_METAMAP r_data_metamap  where r_data_meta_main.meta_attr_name = ?  AND R_DATA_MAIN.data_id BETWEEN ? AND ?  AND r_data_metamap.meta_id = r_data_meta_main.meta_id  AND R_DATA_MAIN.data_id = r_data_metamap.object_id
> Sep 13 11:36:33 pid:6388 NOTICE: SQLSTATE: 08S01
> Sep 13 11:36:33 pid:6388 NOTICE: SQLCODE: 35
> Sep 13 11:36:33 pid:6388 NOTICE: SQL Error message: server closed the connection unexpectedly
>         This probably means the server terminated abnormally
>         before or while processing the request.
> ;
> The connection has been lost
> ```

By the hour offset to failure & "the server end" being the same host, I infer that times above are UTC.

(I'm running the performance test on the iCAT server host, as the `irods` user. We don't share the host or the user in production.)


On this first failure our DBA said

> In every case where we have seen this message recently ...
> 
> > Sep 13 11:36:33 pid:6388 NOTICE: SQL Error message: server closed the connection unexpectedly  
> >        This probably means the server terminated abnormally  
> >        before or while processing the request.
> 
> ... the problem has been at the client end, caused by one kind of Openstack networking issue or another.  The database has been running normally for almost a month, and the log file contains no message indicating that it closed any client connections this morning for any reason.

**Comment by trel** [comment]
From the DBA.... the 'server' is the database server, and the 'client' is the iRODS Agent running as a fork of the iRODS Server... 

However, the iRODS log says "SQL Error message: server closed the connection unexpectedly" which suggests the database server itself dropped the connection.

Am I interpreting correctly that the two ends of the database connection are 'blaming' each other?   Or rather... to grok what you're saying about the OpenStack networking... if the network hiccups for a split second... both ends are blaming the middle?

If that's the case, and it's so very intermittent... is there anything we can do to help with this?   Always open for suggestions.  Slightly different ODBC settings/knobs?   More logging of something?

**Comment by mcast** [comment]
On Wed, Sep 15, 2021 at 05:23:09AM -0700, Terrell Russell wrote:

> From the DBA.... the 'server' is the database server, and the
> 'client' is the iRODS Agent running as a fork of the iRODS Server...

Yes, that's my understanding.

> However, the iRODS log says "SQL Error message: server closed the
> connection unexpectedly" which suggests the database server itself
> dropped the connection.
> 
> Am I interpreting correctly that the two ends of the database
> connection are 'blaming' each other?  Or rather... to grok what
> you're saying about the OpenStack networking... if the network
> hiccups for a split second... both ends are blaming the middle?

It does seem that both ends are blaming the middle, but at different
times.  And we do have to remember that there have been works on
Openstack infrastructure (networking and hypervisors) recently, but we
haven't brought in the sysads doing that because we think it happened
at different times.

How long should the iRODS Agent be waiting on a connection which turns
out to be dead?

The postgres server is configured without timeouts (below).


> If that's the case, and it's so very intermittent... is there
> anything we can do to help with this?  Always open for suggestions.
> Slightly different ODBC settings/knobs?  More logging of something?

I was hoping you might be able to advise on iRODS Agent <--> Postgres
timeouts; or log settings we could tweak.

Or that the timings of the assorted reports of the one failure might
spark some suspicion?


Failing that the ticket is FYI (you plural, including future readers),
and so if it become a problem in production we have history from which
to escalate investigations.


Our DBA wrote:

> We do not configure any timeouts for idle connections or long queries in any of our PostgreSQL clusters.  The two parameters that might possibly be relevant have their default value of zero:
>
> ```
> dba:postgres =# select current_setting('statement_timeout');
>  current_setting
>  -----------------
>   0
>   (1 row)
> 
> dba:postgres =# select current_setting('idle_in_transaction_session_timeout');
>  current_setting
>  -----------------
>   0
> ```
> 
> And in this context, as the documentation
> 
>   https://www.postgresql.org/docs/11/runtime-config-client.html#RUNTIME-CONFIG-CLIENT-STATEMENT
> 
> makes clear, zero means that timeouts are disabled.
> 
> I hope this clarifies the situation at the database end.


Keep well,
-- 
Matthew


-- 
 The Wellcome Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by kript** [comment]
One of my troubleshooting steps for this was to start an `imeta` session on the Provider and then leave it running.
It was still up the next morning, which might rule out a database timeout, and probably an OS connection timeout, as the connection could be seen on the database server even though no commands were being run.

**Comment by mcast** [comment]
(for internal readers: https://gitlab.internal.sanger.ac.uk/casm/cgpDataOut/-/issues/271#note_116886 )

**Comment by mcast** [comment]
Continuing from
> More about the first failure,
> 
> ```
> :run(2021-09-13 09:35:14.292242) iquest  --no-page  -z  /cgp  %s %s %s  select min(DATA_ID), max(DATA_ID), count(DATA_ID) where META_DATA_ATTR_NA
> ME = 'id_ifile' and DATA_ID between '37041408' '51693660'; [7279.382s]                                                                                         
> ERROR: iquest Error: queryAndShowStrCond failed status = -806000 CAT_SQL_ERR                                                                                
> ```
> 
> Again UTC, so it ran 10:35:14 to 12:36:33 BST.

DBA reports:

> The only entry in the PostgreSQL cluster log for 13 September is this one:
> 
> ```
> 2021-09-13 10:51:03.533 BST [1796764] irods@irodscgp LOG:  could not receive data from client: Connection timed out
> ```

The postgres server noticed a problem after 15m 49sec, but the client (iRODS Agent) noticed after 7279 sec = 2:01:19.

Suppose the query itself took 19 seconds (which is marginally slower than 99% of its siblings).  Server noticed after 15:30 and client after 2:01:00, which starts to look like the sort of round numbers a programmer might pick for some kinds of things.

We're looking at what else we can do to find out more...

**Comment by mcast** [comment]
I wondered what the size of the answers to those two `iquest`s was,

* First (13th Sep)
    * `select min(DATA_ID), max(DATA_ID), count(DATA_ID) where META_DATA_ATTR_NAME = 'id_ifile' and DATA_ID between '37041408' '51693660'`
    * 2 lines, 43 bytes
    * ```
      Zone is /cgp
      37041410..51693660..9038594..
      ```
* Second (15th Sep)
    * `iquest  --no-page  -z  /cgp  "%s..%s..%s..%s..%s..%s..%s..%s..%s..%s.." "select ORDER(DATA_ID), ORDER(DATA_REPL_NUM), DATA_REPL_STATUS, COLL_NAME, DATA_NAME, DATA_CHECKSUM, DATA_SIZE, DATA_RESC_HIER, DATA_PATH, META_DATA_ATTR_VALUE where META_DATA_ATTR_NAME = 'id_ifile' and DATA_ID between '23876440' '23990845'" | wc`
    * `226753  226755 58647373` (56 MiB)

and from some iRODS logs the phrase `chlGenQuery cmlGetFirstRowFromSql` makes me wonder whether the failure happened before or after data was shovelled to the other end..?  Sounds like the failure was before the data on 15th Sep.

(`..` is not the usual field separator, but it's easier to read if I change them here)

**Comment by trel** [comment]
We'll work to write up a state machine / explanation and get that into this thread.


>- For the iRODS Agent to iCAT backend database, in Postgres mode, is there any connection pooling or re-use?  My observations suggest that connections are made on-demand, used for the duration of an `iquest` and then closed which is nice and simple...  but I'm only watching `netstat -t` here.

Correct - there is no database connection pooling in the iRODS Server (https://github.com/irods/irods/issues/2063).

>- When `CAT_SQL_ERR` occurs, what clues could we extract from the logs about the state of the query and its underlying TCP connection?

`CAT_SQL_ERR` usually/has always(?) means/meant that a query was malformed in some way, but in this case, we have determined that is not the case... the query itself is fine... it's just that the connection died for some reason along the way.  We'll think through how to disambiguate this case a bit better - not sure what is available at this time without that better disambiguation/logging.

>- What else we can do to discover the state(s) of the machine(s)?
>- Other than repeatedly querying and hoping for failure, what would you suggest might provoke this kind of problem?

Potentially turn up the logging, but that gets very noisy, and we don't have a surgical test case to trip this event.  We may be able to write a test that severs the connection / drops the packets... will discuss internally.  We have https://github.com/irods/mungefs for messing with filesystem access... will look for some network fault injection tooling we can incorporate.

**Comment by trel** [comment]
We have experience with `tc` already for 'testing at a distance', aka introducing delay - but it can also introduce packet loss, so this seems a way forward to reliably reproduce this 'incorrect' `CAT_SQL_ERR`.

**Comment by trel** [comment]
`tc` may not be sufficient - rather `iptables` could be all that is needed...

https://stackoverflow.com/a/1088475/6561189

**Comment by mcast** [comment]
Looks like success! An extra Openstack migration for testing this problem broke all four of the running `iquest`s I had going. Sysad reports

> I lost network connectivity to it for 4.8s

They failed with `CAT_SQL_ERR` in (1270.200 1274.016 1275.907 1278.768) seconds.
* I interpret this as ( `net.ipv4.tcp_keepalive_time` = 1200, plus one minute of unknown origin, plus X ).
    * We changed `net.ipv4.tcp_keepalive_time` from 7200 after the first two failures.
    * I have not finished reading https://tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/
* I then infer X to be either the length of time `iquest` took to do most of the job, having got everything it needed, but then failing to shut down the SQL connection cleanly; or almost zero when the SQL had not started before network was disrupted.

What still puzzles me is that my ssh connections stayed up. They're going in the opposite direction (to port 22 on the iRODS host), but I believe this isn't relevant for TCP timeout.

----

I think this is good news, but we're still discussing it.

**Comment by trel** [comment]
Excellent.   This does appear to make sense.

If you can confirm that those `iquest` queries were, in fact.. 10, 14, 15, 18? seconds long to return from the db itself ... then I think you've done it.  I don't see the 'zero' seconds you are talking about if 1200+60=1260 and the first `iquest` failed at 1270.   But it seems boxed in.   Now, we need to figure out how to reproduce this without OpenStack and we can write a test and update the iRODS error message perhaps to be more helpful.

I think SSH has its own `SSH keepalive packets`... so more/different knobs for that connection.

**Comment by kript** [comment]
Something its highlighted for us is our lack of understanding of how the database connection layer/library works. 
Obviously #2063 would help, but so would detecting disconnection and reconnecting - which is something we/our DBA's always recommend out developers code for defensively - handle the database going away and reconnect.
This is likely another issue/enhancement request, but I thought I would mention here in case there are any features of the underlying library you could expose.

**Comment by trel** [comment]
Pooling presents its own issues...

for this discussion...

db_open_op - https://github.com/irods/irods/blob/4.2.10/plugins/database/src/db_plugin.cpp#L1887
calls
cmlOpen - https://github.com/irods/irods/blob/4.2.10/plugins/database/src/mid_level_routines.cpp#L91
calls
cllConnect - https://github.com/irods/irods/blob/4.2.10/plugins/database/src/low_level_odbc.cpp#L169
all communicating via `icatSessionStruct *icss`

`CAT_SQL_ERR` is returned in eight different cases in...
https://github.com/irods/irods/blob/4.2.10/plugins/database/src/mid_level_routines.cpp
but from only three unique low-level calls...
- cllExecSqlNoResult
- cllExecSqlWithResultBV
- cllExecSqlWithResult

The thinking is that we could check error codes a bit more specifically, instead of just passing back `CAT_SQL_ERR`.   But otherwise, we'd have to potentially engineer a new connect/reconnect logic/approach.

**Comment by mcast** [comment]
> I've asked for advice on what may have happened to packets. Possibly they're dropped, or a NAT forgets the connection and continues dropping packets after?

Sorry this is the best information we have:

> during the 4.8s there would have been no packets transmitted / received - TCP should resend, and, as the SSH sessions to the instance were not broken (they would have paused for a few seconds) then it demonstrates that the TCP sessions were preserved through the transition. I'm not entirely sure why this pause should have impacted the higher level protocols

**Comment by mcast** [comment]
I'm now poking around with this briefly for my own education...  probably you already knew this but I'm catching up.

Noting that in https://tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/ it says
> Remember that keepalive support, even if configured in the kernel, is not the default behavior in Linux. Programs must request keepalive control for their sockets using the setsockopt interface. There are relatively few programs implementing keepalive, but you can easily add keepalive support for most of them following the instructions explained later in this document. 

## ssh uses keepalive
```
$ strace -T -tt  -f -s 8192 -o mca.ssh.strace ssh localhost
...
$ grep setsockopt mca.ssh.strace 
384   09:39:21.279604 setsockopt(3, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0 <0.000009>
```

I didn't verify that it is specifically the ssh interactive session getting the option, because I don't think there are any others.

## iRODS Agent's postgres client does too
```
$ sudo strace -T -tt  -ff -s 8192 -o mca.irods.strace -p 2686 -p 25598 -p 25599 -p 25600
...
```
(run an iquest, lots of stuff is traced)

Find one making a postgres connection - there were several, I just picked the last.
```
09:48:43.810375 socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 10 <0.000011>
09:48:43.810402 setsockopt(10, SOL_TCP, TCP_NODELAY, [1], 4) = 0 <0.000006>
09:48:43.810427 fcntl(10, F_GETFL)      = 0x2 (flags O_RDWR) <0.000005>
09:48:43.810447 fcntl(10, F_SETFL, O_RDWR|O_NONBLOCK) = 0 <0.000005>
09:48:43.810466 fcntl(10, F_SETFD, FD_CLOEXEC) = 0 <0.000005>
09:48:43.810485 setsockopt(10, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0 <0.000006>
09:48:43.810507 connect(10, {sa_family=AF_INET, sin_port=htons(5432), sin_addr=inet_addr("172.17.154.48")}, 16) = -1 EINPROGRESS (Operation now in progress) <0.000034>
```

and that's the connection being made, with `SO_KEEPALIVE` set.

**Comment by mcast** [comment]
On both (iCAT server / postgres client) and (postgres server), we have `tail /proc/sys/net/ipv4/tcp_keepalive_*` saying
```
==> /proc/sys/net/ipv4/tcp_keepalive_intvl <==
75

==> /proc/sys/net/ipv4/tcp_keepalive_probes <==
9

==> /proc/sys/net/ipv4/tcp_keepalive_time <==
1200
```
